tensorflow/python/kernel_tests/xent_op_deterministic_test.py

```diff
@@ -94,39 +73,132 @@ def _generateInputs(self, dtype, seed=123):
 
   @test_util.run_in_graph_and_eager_modes
   def testForward(self):
-    with self.session(), test_util.force_cpu():
-      for dtype in [np.float16, np.float32, np.float64]:
-        for trial in range(5):
-          seed = 123 + trial
-          labels, logits = self._generateInputs(dtype, seed=seed)
-          result_a = nn_ops.softmax_cross_entropy_with_logits_v2(
-              labels=labels, logits=logits)
-          result_b = nn_ops.softmax_cross_entropy_with_logits_v2(
-              labels=labels, logits=logits)
-          self.assertAllEqual(result_a, result_b)
+    for use_gpu in [False, True]:
+      if use_gpu and self._gpuNotAvailable():
+        continue
+      with self.cached_session(), test_util.device(use_gpu):
+        # Even in eager mode, the above line will be able to pin ops to CPU.
+        for dtype in [np.float16, np.float32, np.float64]:
+          for trial in range(5):
+            seed = 123 + trial
+            labels, logits = self._generateInputs(
+                dtype, seed=seed, forward_not_backward=True)
+            result_a = nn_ops.softmax_cross_entropy_with_logits(
+                labels=labels, logits=logits)
+            result_b = nn_ops.softmax_cross_entropy_with_logits(
+                labels=labels, logits=logits)
+            self.assertAllEqual(result_a, result_b)
+
+  def _testBackward(self, labels_not_logits=False):
+    for use_gpu in [False, True]:
+      if use_gpu and self._gpuNotAvailable():
+        continue
+      with self.cached_session(), test_util.device(use_gpu):
+        # Even in eager mode, the above line will be able to pin ops to CPU.
+        for dtype in [np.float16, np.float32, np.float64]:
+          labels, logits = self._generateInputs(dtype, seed=456)
+          output_shape = labels.shape[0]
+
+          def gradients(seed=789):
+            np.random.seed(seed)
+            upstream_gradients = self._randomFloats(output_shape, dtype)
+            with backprop.GradientTape(persistent=True) as tape:
+              tape.watch(labels)
+              tape.watch(logits)
+              op_output = nn_ops.softmax_cross_entropy_with_logits(
+                  labels=labels, logits=logits)
+              gradient_injector_output = op_output * upstream_gradients
+            return tape.gradient(gradient_injector_output, [labels, logits])
+
+          for trial in range(5):
+            seed = 456 + trial
+            labels_grad_a, logits_grad_a = gradients(seed=seed)
+            labels_grad_b, logits_grad_b = gradients(seed=seed)
+            if labels_not_logits:
```

Comment:
Instead of having this parameter, i would just assert both labels and logits.

Reply:
Right. That's how this test started out. It evolved so that each op-path is covered independently (with code re-use but less efficient use of compute cycles). For one thing, this made it easy for me to confirm (without modification to the tests) that the tests will catch nondeterminism in each path independently. With these tests, for example, I disabled the determinism solution and confirmed that both `testBackwardLogits` and `testBackwardLabels` failed. I believe that tests quit on the first assert, which makes it impossible to fully validate their intention without modifying them (albeit, just the addition or subtraction of a `#`).

I'm considering simplifying this.

Reply:
The next commit will remove this complexity.

Reply:
If you want to test multiple things independently (allowing the second to run even if the first fails), consider using [`self.subtest`](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests). I don't think it's necessary in this case though.